runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/DoFnOp.java

```diff
@@ -237,9 +292,39 @@ private String getTimerStateId(DoFnSignature signature) {
     return builder.toString();
   }
 
+  private void attemptStartBundle() {
+    if (isBundleStarted.compareAndSet(false, true)) {
```

Comment:
thread safety is still bit awkward here as if another thread invokes attemptStartBundle() in a different thread after the the line of finishBundle, the count and time can still be set in a non-consistent manner. I guess we don't allow multiple threads in the same task right now. But it can cause problems. Could you add a check in the translator to make sure we don't have task.max.concurrency bigger than 1 if bundling is enabled?

Reply:
Make sense. Will do